redirect/src/main/java/com/adyen/checkout/redirect/RedirectUtil.kt

```diff
@@ -107,28 +84,98 @@ object RedirectUtil {
      * @return And intent that targets either another app or a Web page.
      */
     @JvmStatic
-    fun createRedirectIntent(context: Context, uri: Uri): Intent {
-        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
-            createCustomTabIntent(context, uri)
-        } else {
-            if (determineResolveResult(context, uri).type == ResolveResult.Type.APPLICATION) {
-                Intent(Intent.ACTION_VIEW, uri)
-            } else {
-                createCustomTabIntent(context, uri)
+    fun launchUri(context: Context, uri: Uri) {
+        val launched = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R)
```

Comment:
Sorry for being nitpicky but I have an idea. If we extract another `launchNative` method and we also make `launchBrowser` return a boolean, we can do this:

```kotlin
    fun launchUri(context: Context, uri: Uri) {
        if (launchNative(context, uri)) return
        if (launchWithCustomTabs(context, uri)) return
        if (launchBrowser(context, uri)) return
        throw ComponentException("Redirect to app failed.")
    }

    private fun launchNative(context: Context, uri: Uri): Boolean {
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) launchNativeApi30(context, uri)
        else launchNativeBeforeApi30(context, uri)
    }
```

What do you think?

Reply:
you're so picky 😂

Reply:
whoops sorry I got a build error for that 
```
ReturnCount - [launchUriRedirect] at /Users/atef/android-projects/adyen-android/redirect/src/main/java/com/adyen/checkout/redirect/RedirectUtil.kt:80:9
```
I'll just do it like this
```
fun launchUriRedirect(context: Context, uri: Uri) {
        val launchNative = launchNative(context, uri)
        if (!launchNative) {
            val launchCustomWithTabs = launchWithCustomTabs(context, uri)
            if (!launchCustomWithTabs)
                launchBrowser(context, uri)
        }
    }
```
```
 /**
     * in case the device doesn't support custom tabs or doesn't support google services (Huawei device).
     */
    private fun launchBrowser(context: Context, uri: Uri) {
        try {
            val browserActivityIntent = Intent()
                .setAction(Intent.ACTION_VIEW)
                .addCategory(Intent.CATEGORY_BROWSABLE)
                .addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                .setData(uri)
            context.startActivity(browserActivityIntent)

        } catch (e: ActivityNotFoundException) {
            throw ComponentException("Redirect to app failed.")
        }
    }
```
let me know what do you think @jreij ?

Reply:
There are 2 points here:
- I'm trying to avoid throwing an exception in `launchBrowser`. While it works like this right now because `launchBrowser` is the last method we call, it might cause issues later if we decide to switch the order and call something else after `launchBrowser`. Which is why I suggested making it return a boolean like the other methods, and throwing the exception in the parent method.
- The `if ... return` syntax i suggested was just to make the code more readable and avoid having nested `if`s and boolean `val`s. Feel free to add `@Suppress("ReturnCount")` above `launchUriRedirect` or keep it as you have it now 🙂

Reply:
good points